TELCODOCS-597:  Adding a local registry to a Single Node OpenShift DU profile (#50904)
openshift/openshift-docs: modules/ztp-configuring-pgt-image-registry.adoc

// Module included in the following assemblies:
//
// scalability_and_performance/ztp-deploying-disconnected.adoc

:_module-type: PROCEDURE
[id="ztp-configuring-pgt-image-registry_{context}"]
= Configuring the image registry using PolicyGenTemplate CRs

> You can use `PolicyGenTemplate` to apply to create the PV and PVC and patch `imageregistry` configuration. Select the appropriate `PolicyGenTemplate` for each `source-cr`. See Additional Resources for more help.

Reviewer: You can use `PolicyGenTemplate` CRs? 

"See Additional Resources for more information."

.Prerequisites

> * You have installed and configured Zero Touch Provisioning (ZTP). For further information on this, see the topic on ZTP in Additional resources.

Reviewer: Use "information about..."

.Procedure

> . Add the following YAML to the `StorageClass.yaml` file to configure the image registry:

Reviewer 2: This is incorrect. The `StorageClass.yaml` is the name of the CR file that the you add to the PGT.
 
```suggestion
. Configure the storage class, persistent volume claim, persistent volume, and image registry configuration in the appropriate `PolicyGenTemplate` CR. For example, to configure an individual site, use the following YAML:
```

Author: Done.

+
[source,yaml]
----
sourceFiles:

>   # storage class

Reviewer 3: one more space to align this with the rest

Author: Comments all 3-space aligned.

>    # storage class

Author: Comments aligned.

   - fileName: StorageClass.yaml
     policyName: "sc-for-image-registry"
     metadata:
       name: image-registry-sc
       annotations:
         ran.openshift.io/ztp-deploy-wave: "100" <1>
  # persistent volume claim

>     fileName: StoragePVC.yaml

Reviewer 3: missing `-`

Author: Fixed.

    policyName: "pvc-for-image-registry"
    metadata:
      name: image-registry-pvc
      namespace: openshift-image-registry
        annotations:
          ran.openshift.io/ztp-deploy-wave: "100"  <2>
    spec:
      accessModes:
        ReadWriteMany
      resources:
        requests:
          storage: 100Gi
      storageClassName: image-registry-sc
      volumeMode: Filesystem
   # persistent volume
  - fileName: ImageRegistryPV.yaml <3>
    policyName: "pv-for-image-registry"
    metadata:
      annotations:
        ran.openshift.io/ztp-deploy-wave: "100"  <4>

>    #

Reviewer 3: add `# image registry config`

  - fileName: ImageRegistryConfig.yaml <5>
    policyName: "config-for-image-registry"
    complianceType: musthave <5>
    metadata:
      annotations:
        ran.openshift.io/ztp-deploy-wave: "100" <6>
    spec:
      storage:
        pvc:
          claim: "image-registry-pvc"
----

> <1> Remove this when moved to the correct PGT (`site/group/common`).

Reviewer 2: This is confusing. The YAML example sets `ztp-deploy-wave` to 100 because it is an example of applying the PGT to a single cluster only. You could say for example, 

```suggestion
<1> Set the appropriate value for `ztp-deploy-wave` depending on whether you are configuring image registries at the site, common, or group level. `ztp-deploy-wave: "100"` is appropriate for an individual site.
```

You could also find out what this ztp-deploy-wave value should be when deployed at the group or common level.

Author: We added a link to policy waves documentation in Additional resources.

> <1> Remove this when moved to the correct PGT (`site/group/common`).

Reviewer: Remove what? The value, the entire line? Same for <2> <4> <6>

Define PGT, I had to look in another file above. 

"Remove this (line,value, annotation) when moved to the correct `PolicyGenTemplate` (PGT) ..."

You also mix the language in <4> and <6> : "..when moved to the correct/right PGT". I would use "correct".

<2> Remove this when moved to the correct PGT (`site/group/common`).
<3> This assumes that `mount_point` is set to `/var/imageregistry` in `SiteConfig` using StorageClass `image-registry-sc` (see the topic on configuring disk partitioning with `SiteConfig`).
<4> Remove this when moved to the right PGT (`site/group/common`).
<5> Configure registry to point to the PVC created above.
<6> Remove this when moved to the right PGT (`site/group/common`)

> == Verifying the installation

Reviewer: One heading per module. This needs to be `.Verification` like you have in `.Prerequisites` and `.Procedure`.

> . Check that the CRD Config of the group `imageregistry.operator.openshift.io` instance is not reporting any errors.

Reviewer 3: Config is name -- so `Config` (highlighted)

Author: Done.

> . A few minutes after the installation process is complete, you should see the `PersistentVolumeClaim` filling up.

Reviewer: should   verb
Whenever possible, rewrite in a more direct way. Do not use "should" to refer to an action that must be performed. For example, when parameters must be reset, write "Reset the parameters" and not "You should reset the parameters".

"After installation is complete, the `PeristentVolumeClaim` is in use."

. Check that the `registry*` pod is up correctly located under the `openshift-image-registry` namespace.

From inside the node:

. Verify successful login to the registry with `podman`:
+
[source,yaml]
----

>   oc login -u kubeadmin -p <password_from_install_log> https://api-int.<cluster_name>.<base_domain>:6443

Reviewer: A few things here. 

You have `source,yaml` in what is terminal commands (`source,terminal`). 
Missing a `$` at the start of both commands.
One command per code block.

  podman login -u kubeadmin -p $(oc whoami -t) image-registry.openshift-image-registry.svc:5000
----

. Check for disk partitioning using `lsblk` to list your blocks:
+

> [source,yaml]

Reviewer: source,terminal

----

> [core@mysno ~]$ lsblk

Reviewer 2: `core@mysno` suggests you have SSHed into a cluster, which is fine for development purposes, but not ideal for customer facing docs, since SSH is not recommended for production cluster nodes. Suggest using `oc/debug node` as outlined here: https://github.com/openshift/openshift-docs/pull/53083/files#diff-a8c623d5afaf11e6c3cbbd690e508160a70fdeb8cd149a0d92bccb067206efd2R129

Author: Resolved.

> NAME   MAJ:MIN RM   SIZE RO TYPE MOUNTPOINT

Reviewer: Separate the output from the command and remove the `[core@mysno ~]`. 

Please review [Code blocks, command syntax, and example output](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#code-blocks-command-syntax-and-example-output)

sda      8:0    0 446.6G  0 disk
  |-sda1   8:1    0     1M  0 part
  |-sda2   8:2    0   127M  0 part
  |-sda3   8:3    0   384M  0 part /boot
  |-sda4   8:4    0 336.3G  0 part /sysroot
  `-sda5   8:5    0 100.1G  0 part /var/imageregistry <1>
sdb      8:16   0 446.6G  0 disk
sr0     11:0    1   104M  0 rom
----

> <1> This setting will appear if you have successfully listed your block.

Reviewer 3: remove `<1>`

Reviewer 3: oh never mind about this one -- it's pointing to console output

// Module included in the following assemblies:
//
// scalability_and_performance/ztp-deploying-disconnected.adoc

:_module-type: PROCEDURE
[id="ztp-configuring-pgt-image-registry_{context}"]
= Configuring the image registry using PolicyGenTemplate CRs

You can use `PolicyGenTemplate` to apply to create the PV and PVC and patch `imageregistry` configuration. Select the appropriate `PolicyGenTemplate` for each `source-cr`. See Additional Resources for more help.

> .Prerequisites

Reviewer 3: Generally....the pre req is that the user has ZTP setup

Author: Added "* You have installed and configured Zero Touch Provisioning (ZTP)."

     metadata:
       name: image-registry-sc
       annotations:
         ran.openshift.io/ztp-deploy-wave: "100" <1>
  # persistent volume claim
  - fileName: StoragePVC.yaml
    policyName: "pvc-for-image-registry"
    metadata:
      name: image-registry-pvc
      namespace: openshift-image-registry
        annotations:
          ran.openshift.io/ztp-deploy-wave: "100"  <2>
    spec:
      accessModes:
        - ReadWriteMany
      resources:
        requests:
          storage: 100Gi
      storageClassName: image-registry-sc
      volumeMode: Filesystem
   # persistent volume
  - fileName: ImageRegistryPV.yaml <3>
    policyName: "pv-for-image-registry"
    metadata:
      annotations:
        ran.openshift.io/ztp-deploy-wave: "100"  <4>
   #
  - fileName: ImageRegistryConfig.yaml <5>
    policyName: "config-for-image-registry"
    complianceType: musthave <5>
    metadata:
      annotations:
        ran.openshift.io/ztp-deploy-wave: "100" <6>
    spec:
      storage:
        pvc:
          claim: "image-registry-pvc"
----
<1> Remove this when moved to the correct PGT (`site/group/common`).
<2> Remove this when moved to the correct PGT (`site/group/common`).

> <3> This assumes that `mount_point` is set to `/var/imageregistry` in `SiteConfig` using StorageClass `image-registry-sc` (see the first sc-file).

Reviewer 3: Change "see the first sc-file" to something like See "StieConfig from Step 1"

Author: Changed to (see the topic on configuring disk partitioning with `SiteConfig`)

// Module included in the following assemblies:
//
// scalability_and_performance/ztp-deploying-disconnected.adoc

:_module-type: PROCEDURE
[id="ztp-configuring-pgt-image-registry_{context}"]
= Configuring the image registry using PolicyGenTemplate CRs

> You can configure PTP fast events for vRAN clusters that are deployed using the GitOps Zero Touch Provisioning (ZTP) pipeline. Use `PolicyGenTemplate` custom resources (CRs) as the basis to create a hierarchy of configuration files tailored to your specific site requirements.

Reviewer 4: I don't think `PTP fast events` are related to this section.

Author: Removed.

.Prerequisites

* You have installed and configured Zero Touch Provisioning (ZTP). For further information on this, see the topic on ZTP in Additional resources.

* You configured the hub cluster for generating the required installation and policy CRs.

> * You created a Git repository where you manage your custom site configuration data. The repository must be accessible from the hub cluster and be defined as a source repository for the Argo CD application.

Reviewer 3: If ZTP is setup correctly -- these highlight statements are implied to be already in place

.Procedure

. Add the following YAML to the `StorageClass.yaml` file to configure the image registry:
+
[source,yaml]
----
sourceFiles:
   # storage class
   - fileName: StorageClass.yaml
     policyName: "sc-for-image-registry"
     metadata:
       name: image-registry-sc
       annotations:
         ran.openshift.io/ztp-deploy-wave: "100" <1>
  # persistent volume claim

>   - fileName: StoragePVC.yaml

Reviewer 4: This line doesn't have the same indentation level with line 28

Author: Fixed.

Reviewer: @tmulquee  it needs an additional space. It does not render correctly in your [preview](http://file.emea.redhat.com/tmulquee/TELCODOCS-597/scalability_and_performance/ztp-deploying-disconnected.html#ztp-configuring-pgt-image-registry_ztp-deploying-disconnected).

:_module-type: PROCEDURE
[id="ztp-configuring-pgt-image-registry_{context}"]
= Configuring the image registry using PolicyGenTemplate CRs

You can use `PolicyGenTemplate` to apply to create the PV and PVC and patch `imageregistry` configuration. Select the appropriate `PolicyGenTemplate` for each `source-cr`. See Additional Resources for more help.

.Prerequisites

* You have installed and configured Zero Touch Provisioning (ZTP). For information about this, see the topic on ZTP in Additional resources.

.Procedure

. Configure the storage class, persistent volume claim, persistent volume, and image registry configuration in the appropriate `PolicyGenTemplate` CR. For example, to configure an individual site, use the following YAML:
+
[source,yaml]
----
sourceFiles:
   # storage class
   - fileName: StorageClass.yaml
     policyName: "sc-for-image-registry"
     metadata:
       name: image-registry-sc
       annotations:
         ran.openshift.io/ztp-deploy-wave: "100" <1>
  # persistent volume claim
   - fileName: StoragePVC.yaml
    policyName: "pvc-for-image-registry"
    metadata:
      name: image-registry-pvc
      namespace: openshift-image-registry
        annotations:

>           ran.openshift.io/ztp-deploy-wave: "100"  <2>

Reviewer 2: Do we need the same note repeated at every annotation? I think probably just having the first note is OK.

Author: Done.

<2> Set the appropriate value for `ztp-deploy-wave` as in note 1.
<3> This assumes that `mount_point` is set to `/var/imageregistry` in `SiteConfig` using StorageClass `image-registry-sc` (see the topic on configuring disk partitioning with `SiteConfig`).
<4> Set the appropriate value for `ztp-deploy-wave` as in note 1.
<5> Configure registry to point to the PVC created above.
<6> Set the appropriate value for `ztp-deploy-wave` as in note 1.

.Verification

. Check that the CRD `Config` of the group `imageregistry.operator.openshift.io` instance is not reporting any errors.

. A few minutes after the installation process is complete, you should see the `PersistentVolumeClaim` filling up.

. Check that the `registry*` pod is up correctly located under the `openshift-image-registry` namespace.

From inside the node:

. Verify successful login to the registry with `podman`:
+
[source,terminal]
----

>   oc login -u kubeadmin -p <password_from_install_log> https://api-int.<cluster_name>.<base_domain>:6443

Reviewer 2: This is actually two commands, and they are both missing the $ symbol. What about:

```suggestion
[source,terminal]
----
$ oc login -u kubeadmin -p <password_from_install_log> https://api-int.<cluster_name>.<base_domain>:6443
----
+
[source,terminal]
----
podman login -u kubeadmin -p $(oc whoami -t) image-registry.openshift-image-registry.svc:5000
----
```

Author: Done.

> [source,yaml]

Reviewer 2: ```suggestion
[source,terminal]
```

Author: Done.

  podman login -u kubeadmin -p $(oc whoami -t) image-registry.openshift-image-registry.svc:5000
----

. Check for disk partitioning using `lsblk` to list your blocks:
+

> [source,yaml]

Reviewer 2: ```suggestion
[source,terminal]
```

        ran.openshift.io/ztp-deploy-wave: "100"  <4>
  # image registry config
  - fileName: ImageRegistryConfig.yaml <5>
    policyName: "config-for-image-registry"
    complianceType: musthave <5>
    metadata:
      annotations:
        ran.openshift.io/ztp-deploy-wave: "100" <6>
    spec:
      storage:
        pvc:
          claim: "image-registry-pvc"
----
<1> Set the appropriate value for `ztp-deploy-wave` depending on whether you are configuring image registries at the site, common, or group level. `ztp-deploy-wave: "100"` is appropriate for an individual site.  ZTP deploy waves are used to order how policies are applied to the spoke cluster. All policies created by `PolicyGen` have a ztp deploy wave by default.
<2> Set the appropriate value for `ztp-deploy-wave` as in note 1.
<3> This assumes that `mount_point` is set to `/var/imageregistry` in `SiteConfig` using StorageClass `image-registry-sc` (see the topic on configuring disk partitioning with `SiteConfig`).
<4> Set the appropriate value for `ztp-deploy-wave` as in note 1.
<5> Configure registry to point to the PVC created above.
<6> Set the appropriate value for `ztp-deploy-wave` as in note 1.

.Verification

> . Check that the CRD `Config` of the group `imageregistry.operator.openshift.io` instance is not reporting any errors.

Author: Done.

> . A few minutes after the installation process is complete, you should see the `PersistentVolumeClaim` filling up.

Author: Done.

. Check that the `registry*` pod is up correctly located under the `openshift-image-registry` namespace.

> From inside the node:

Author: Done.

. Verify successful login to the registry with `podman`:
+
[source,terminal]
----
$ oc login -u kubeadmin -p <password_from_install_log> https://api-int.<cluster_name>.<base_domain>:6443
----
+
[source,terminal]
----

> podman login -u kubeadmin -p $(oc whoami -t) image-registry.openshift-image-registry.svc:5000

Author: Done.
